In [2]:
import csv
import pandas as pd
import operator
from xgboost import XGBRegressor # you have to install it first!
import sklearn
from sklearn.preprocessing import Imputer
from sklearn.impute import SimpleImputer

ModuleNotFoundError: No module named 'sklearn.impute'

In [2]:
sample = pd.read_csv(r'all\sample_submission.csv', sep = ',')
sample.head(5)

Id      SalePrice
0  1461  169277.052498
1  1462  187758.393989
2  1463  183583.683570
3  1464  179317.477511
4  1465  150730.079977

# Test file

In [3]:
test = pd.read_csv(r'all\test.csv', sep = ',')
test.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   
5  1466          60       RL         75.0    10000   Pave   NaN      IR1   
6  1467          20       RL          NaN     7980   Pave   NaN      IR1   
7  1468          60       RL         63.0     8402   Pave   NaN      IR1   
8  1469          20       RL         85.0    10176   Pave   NaN      Reg   
9  1470          20       RL         70.0     8400   Pave   NaN      Reg   

  LandContour Utilities      ...       ScreenPorch PoolArea PoolQC  Fence  \
0         Lvl    AllPub      ...               120        0    NaN  MnPrv   
1         Lvl    AllPub      ...                 0        0    NaN    NaN   
2         Lvl    AllPub      ...                 0        0    NaN  MnPrv   
3         Lvl    AllPub      ...                 0        0    NaN    NaN   
4         HLS    AllPub      ...               144        0    NaN    NaN   
5         Lvl    AllPub      ...                 0        0    NaN    NaN   
6         Lvl    AllPub      ...                 0        0    NaN  GdPrv   
7         Lvl    AllPub      ...                 0        0    NaN    NaN   
8         Lvl    AllPub      ...                 0        0    NaN    NaN   
9         Lvl    AllPub      ...                 0        0    NaN  MnPrv   

  MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0      6    2010        WD         Normal  
1        Gar2   12500      6    2010        WD         Normal  
2         NaN       0      3    2010        WD         Normal  
3         NaN       0      6    2010        WD         Normal  
4         NaN       0      1    2010        WD         Normal  
5         NaN       0      4    2010        WD         Normal  
6        Shed     500      3    2010        WD         Normal  
7         NaN       0      5    2010        WD         Normal  
8         NaN       0      2    2010        WD         Normal  
9         NaN       0      4    2010        WD         Normal  

[10 rows x 80 columns]

In [4]:
len(test)

1459

In [5]:
test.describe()

Id   MSSubClass  LotFrontage       LotArea  OverallQual  \
count  1459.000000  1459.000000  1232.000000   1459.000000  1459.000000   
mean   2190.000000    57.378341    68.580357   9819.161069     6.078821   
std     421.321334    42.746880    22.376841   4955.517327     1.436812   
min    1461.000000    20.000000    21.000000   1470.000000     1.000000   
25%    1825.500000    20.000000    58.000000   7391.000000     5.000000   
50%    2190.000000    50.000000    67.000000   9399.000000     6.000000   
75%    2554.500000    70.000000    80.000000  11517.500000     7.000000   
max    2919.000000   190.000000   200.000000  56600.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1459.000000  1459.000000   1459.000000  1444.000000  1458.000000   
mean      5.553804  1971.357779   1983.662783   100.709141   439.203704   
std       1.113740    30.390071     21.130467   177.625900   455.268042   
min       1.000000  1879.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1953.000000   1963.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1992.000000     0.000000   350.500000   
75%       6.000000  2001.000000   2004.000000   164.000000   753.500000   
max       9.000000  2010.000000   2010.000000  1290.000000  4010.000000   

          ...        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch  \
count     ...       1458.000000  1459.000000  1459.000000    1459.000000   
mean      ...        472.768861    93.174777    48.313914      24.243317   
std       ...        217.048611   127.744882    68.883364      67.227765   
min       ...          0.000000     0.000000     0.000000       0.000000   
25%       ...        318.000000     0.000000     0.000000       0.000000   
50%       ...        480.000000     0.000000    28.000000       0.000000   
75%       ...        576.000000   168.000000    72.000000       0.000000   
max       ...       1488.000000  1424.000000   742.000000    1012.000000   

         3SsnPorch  ScreenPorch     PoolArea       MiscVal       MoSold  \
count  1459.000000  1459.000000  1459.000000   1459.000000  1459.000000   
mean      1.794380    17.064428     1.744345     58.167923     6.104181   
std      20.207842    56.609763    30.491646    630.806978     2.722432   
min       0.000000     0.000000     0.000000      0.000000     1.000000   
25%       0.000000     0.000000     0.000000      0.000000     4.000000   
50%       0.000000     0.000000     0.000000      0.000000     6.000000   
75%       0.000000     0.000000     0.000000      0.000000     8.000000   
max     360.000000   576.000000   800.000000  17000.000000    12.000000   

            YrSold  
count  1459.000000  
mean   2007.769705  
std       1.301740  
min    2006.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2010.000000  

[8 rows x 37 columns]

In [6]:
# set of columns with at least one NaN value
nan_dict = {}
for col in test:
    nan_count = test[test[col].isna()]
    if len(nan_count)>0:
        nan_dict[col] = len(nan_count)
        
sorted_nan = sorted(nan_dict.items(), key = operator.itemgetter(1))

In [7]:
sorted_nan.reverse()
sorted_nan

[('PoolQC', 1456),
 ('MiscFeature', 1408),
 ('Alley', 1352),
 ('Fence', 1169),
 ('FireplaceQu', 730),
 ('LotFrontage', 227),
 ('GarageCond', 78),
 ('GarageQual', 78),
 ('GarageFinish', 78),
 ('GarageYrBlt', 78),
 ('GarageType', 76),
 ('BsmtCond', 45),
 ('BsmtExposure', 44),
 ('BsmtQual', 44),
 ('BsmtFinType2', 42),
 ('BsmtFinType1', 42),
 ('MasVnrType', 16),
 ('MasVnrArea', 15),
 ('MSZoning', 4),
 ('Functional', 2),
 ('BsmtHalfBath', 2),
 ('BsmtFullBath', 2),
 ('Utilities', 2),
 ('SaleType', 1),
 ('GarageArea', 1),
 ('GarageCars', 1),
 ('KitchenQual', 1),
 ('TotalBsmtSF', 1),
 ('BsmtUnfSF', 1),
 ('BsmtFinSF2', 1),
 ('BsmtFinSF1', 1),
 ('Exterior2nd', 1),
 ('Exterior1st', 1)]

In [8]:
# creating a set of columns names (with NaN values in test)
test_li = set()
for i in range(len(sorted_nan)):
    test_li.add(sorted_nan[i][0])
test_li

{'Alley',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtFinType1',
 'BsmtFinType2',
 'BsmtFullBath',
 'BsmtHalfBath',
 'BsmtQual',
 'BsmtUnfSF',
 'Exterior1st',
 'Exterior2nd',
 'Fence',
 'FireplaceQu',
 'Functional',
 'GarageArea',
 'GarageCars',
 'GarageCond',
 'GarageFinish',
 'GarageQual',
 'GarageType',
 'GarageYrBlt',
 'KitchenQual',
 'LotFrontage',
 'MSZoning',
 'MasVnrArea',
 'MasVnrType',
 'MiscFeature',
 'PoolQC',
 'SaleType',
 'TotalBsmtSF',
 'Utilities'}

**Alley** has 107 null values and the remaining are NaN.  ---> drop ?

**MiscFeature** has 3 second garage, 2 other, 46 shed and the remaining 1408 are missing. We can drop it since we don't need these informations. 

**MiscVal** has 51 different values and the remaining 1408 are missing (*except* miscval = 17000, miscfeature = nan).

**PoolArea** has 1453 zeroes and the remaining 6 are different values. 

**PoolQC** has 2 excellent and 1 good values, the remaining 3 are NaN. The information is not reliable.  ---> drop

In [9]:
col = test['PoolArea']
grouped = col.groupby(col).count()
grouped

PoolArea
0      1453
144       1
228       1
368       1
444       1
561       1
800       1
Name: PoolArea, dtype: int64

In [10]:
# returns the dataframe rows which satisfy the condition 
test.loc[test['MiscVal'] > 0].head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1    1462          20       RL         81.0    14267   Pave   NaN      IR1   
6    1467          20       RL          NaN     7980   Pave   NaN      IR1   
56   1517          85       RL         75.0    11380   Pave   NaN      IR1   
58   1519          20       RL         70.0     8197   Pave   NaN      Reg   
108  1569          50       RL        124.0    18600   Pave   NaN      Reg   
133  1594          90       RH         60.0     7200   Pave  Pave      Reg   
145  1606          80       RL         74.0     9620   Pave   NaN      Reg   
159  1620          90       RL         70.0     7700   Pave   NaN      Reg   
168  1629          80       RL         60.0     8400   Pave   NaN      Reg   
199  1660          20       RL          NaN    10710   Pave   NaN      Reg   

    LandContour Utilities      ...       ScreenPorch PoolArea PoolQC  Fence  \
1           Lvl    AllPub      ...                 0        0    NaN    NaN   
6           Lvl    AllPub      ...                 0        0    NaN  GdPrv   
56          Lvl    AllPub      ...                 0        0    NaN    NaN   
58          Lvl    AllPub      ...                 0        0    NaN  MnPrv   
108         Lvl    AllPub      ...                 0        0    NaN    NaN   
133         Lvl    AllPub      ...                 0        0    NaN    NaN   
145         Lvl    AllPub      ...                 0        0    NaN  GdPrv   
159         Lvl    AllPub      ...                 0        0    NaN    NaN   
168         Lvl    AllPub      ...                 0        0    NaN    NaN   
199         Lvl    AllPub      ...                 0        0    NaN    NaN   

    MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
1          Gar2   12500      6    2010        WD         Normal  
6          Shed     500      3    2010        WD         Normal  
56         Shed    1500      1    2010        WD         Normal  
58         Shed     300      2    2010        WD         Normal  
108        Shed     450      6    2010        WD         Normal  
133        Shed     500      2    2010        WD         Normal  
145        Shed      80      5    2010        WD         Normal  
159        Shed     600      2    2010        WD         Normal  
168        Shed     490      6    2009        WD         Normal  
199        Shed    1200      7    2009        WD         Normal  

[10 rows x 80 columns]

# Train

In [11]:
train = pd.read_csv(r'all\train.csv', sep = ',')
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

In [12]:
train.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

           ...         WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count      ...        1460.000000  1460.000000    1460.000000  1460.000000   
mean       ...          94.244521    46.660274      21.954110     3.409589   
std        ...         125.338794    66.256028      61.119149    29.317331   
min        ...           0.000000     0.000000       0.000000     0.000000   
25%        ...           0.000000     0.000000       0.000000     0.000000   
50%        ...           0.000000    25.000000       0.000000     0.000000   
75%        ...         168.000000    68.000000       0.000000     0.000000   
max        ...         857.000000   547.000000     552.000000   508.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  \
count  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000   
mean     15.060959     2.758904     43.489041     6.321918  2007.815753   
std      55.757415    40.177307    496.123024     2.703626     1.328095   
min       0.000000     0.000000      0.000000     1.000000  2006.000000   
25%       0.000000     0.000000      0.000000     5.000000  2007.000000   
50%       0.000000     0.000000      0.000000     6.000000  2008.000000   
75%       0.000000     0.000000      0.000000     8.000000  2009.000000   
max     480.000000   738.000000  15500.000000    12.000000  2010.000000   

           SalePrice  
count    1460.000000  
mean   180921.195890  
std     79442.502883  
min     34900.000000  
25%    129975.000000  
50%    163000.000000  
75%    214000.000000  
max    755000.000000  

[8 rows x 38 columns]

In [13]:
# set of columns with at least one NaN value
nan_dict_train = {}
for col in train:
    nan_count_train = train[train[col].isna()]
    if len(nan_count_train)>0:
        nan_dict_train[col] = len(nan_count_train)
        
sorted_nan_train = sorted(nan_dict_train.items(), key = operator.itemgetter(1))

In [14]:
sorted_nan_train.reverse()
sorted_nan_train

[('PoolQC', 1453),
 ('MiscFeature', 1406),
 ('Alley', 1369),
 ('Fence', 1179),
 ('FireplaceQu', 690),
 ('LotFrontage', 259),
 ('GarageCond', 81),
 ('GarageQual', 81),
 ('GarageFinish', 81),
 ('GarageYrBlt', 81),
 ('GarageType', 81),
 ('BsmtFinType2', 38),
 ('BsmtExposure', 38),
 ('BsmtFinType1', 37),
 ('BsmtCond', 37),
 ('BsmtQual', 37),
 ('MasVnrArea', 8),
 ('MasVnrType', 8),
 ('Electrical', 1)]

In [15]:
# creating a set of columns names (with NaN values in train)
train_li = set()
for i in range(len(sorted_nan_train)):
    train_li.add(sorted_nan[i][0])
train_li

{'Alley',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'BsmtQual',
 'Fence',
 'FireplaceQu',
 'GarageCond',
 'GarageFinish',
 'GarageQual',
 'GarageType',
 'GarageYrBlt',
 'LotFrontage',
 'MSZoning',
 'MasVnrArea',
 'MasVnrType',
 'MiscFeature',
 'PoolQC'}

In [16]:
# all the categories (19) in train_li are also in test_li (33)
print(train_li - test_li)

set()


## An Extension To Imputation

In [20]:
sorted_nan

[('PoolQC', 1456),
 ('MiscFeature', 1408),
 ('Alley', 1352),
 ('Fence', 1169),
 ('FireplaceQu', 730),
 ('LotFrontage', 227),
 ('GarageCond', 78),
 ('GarageQual', 78),
 ('GarageFinish', 78),
 ('GarageYrBlt', 78),
 ('GarageType', 76),
 ('BsmtCond', 45),
 ('BsmtExposure', 44),
 ('BsmtQual', 44),
 ('BsmtFinType2', 42),
 ('BsmtFinType1', 42),
 ('MasVnrType', 16),
 ('MasVnrArea', 15),
 ('MSZoning', 4),
 ('Functional', 2),
 ('BsmtHalfBath', 2),
 ('BsmtFullBath', 2),
 ('Utilities', 2),
 ('SaleType', 1),
 ('GarageArea', 1),
 ('GarageCars', 1),
 ('KitchenQual', 1),
 ('TotalBsmtSF', 1),
 ('BsmtUnfSF', 1),
 ('BsmtFinSF2', 1),
 ('BsmtFinSF1', 1),
 ('Exterior2nd', 1),
 ('Exterior1st', 1)]

In [29]:
drop_test_col = [tup[0] for tup in sorted_nan[:6]]
drop_test_col

['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu', 'LotFrontage']

In [31]:
imp_test_col = [tup[0] for tup in sorted_nan[6:]]
imp_test_col

['GarageCond',
 'GarageQual',
 'GarageFinish',
 'GarageYrBlt',
 'GarageType',
 'BsmtCond',
 'BsmtExposure',
 'BsmtQual',
 'BsmtFinType2',
 'BsmtFinType1',
 'MasVnrType',
 'MasVnrArea',
 'MSZoning',
 'Functional',
 'BsmtHalfBath',
 'BsmtFullBath',
 'Utilities',
 'SaleType',
 'GarageArea',
 'GarageCars',
 'KitchenQual',
 'TotalBsmtSF',
 'BsmtUnfSF',
 'BsmtFinSF2',
 'BsmtFinSF1',
 'Exterior2nd',
 'Exterior1st']

In [34]:
dropped_test = test.drop(drop_test_col, axis = 1)
dropped_train = train.drop(drop_test_col, axis = 1)

In [41]:
# make copy to avoid changing original data (when Imputing)
new_test = dropped_test.copy()

for col in imp_test_col:
    new_test[col + '_was_missing'] = new_test[col].isnull()

# Imputation
my_imputer = Imputer()
new_test = pd.DataFrame(my_imputer.fit_transform(new_test))
new_test.columns = dropped_test.columns

ValueError: could not convert string to float: 'Normal'

In [ ]:
# make copy to avoid changing original data (when Imputing)
new_train = dropped_train.copy()

for col in imp_train_col:
    new_train[col + '_was_missing'] = new_train[col].isnull()

# Imputation
my_imputer = SimpleImputer()
new_train = pd.DataFrame(my_imputer.fit_transform(new_train))
new_train.columns = dropped_train.columns

# Modelling

In [ ]:
reduced_test = test.drop(test_li, axis = 1)
reduced_test.head()

In [ ]:
reduced_train = train.drop(test_li, axis = 1)
reduced_train = reduced_train.drop('SalePrice', axis = 1)
reduced_train.head()

In [ ]:
# converting the categorical data of test
test_predictors = pd.get_dummies(reduced_test)
test_predictors.head()

In [ ]:
# converting the categorical data of train
train_predictors = pd.get_dummies(reduced_train)
train_predictors.head()

In [ ]:
# to ensure the test data is encoded in the same manner
final_train_X, final_test_X = train_predictors.align(test_predictors,
                                                 join = 'left', axis = 1)

In [ ]:
# in order to isolate the price
final_train_y = train.SalePrice

In [ ]:
len(final_train_X)

In [ ]:
my_model = XGBRegressor(n_estimators = 1000, learning_rate=0.05)
my_model.fit(final_train_X, final_train_y)#, early_stopping_rounds = 5)
             #, eval_set = [(test_X, test_y)], verbose = False)

In [ ]:
a = my_model.predict(final_test_X)

#for price in a:
#    print(price)
list(a)

In [ ]:
len(a)

In [ ]:
pred_sale_price = pd.DataFrame(a, columns = ['SalePrice'], index=None)
pred_sale_price['Id'] = test['Id']
pred_sale_price.head()

In [ ]:
cols = pred_sale_price.columns.tolist()
cols = cols[-1:] + cols[:-1]
cols

In [ ]:
pred_sale_price = pred_sale_price[cols]
pred_sale_price.head()

In [ ]:
pred_sale_price.to_csv('predictions.csv', index = False)